Replace toEngine channel with a dedicated subscription API

yacovm opened the pull request:
This commit removes the toEngine channel pattern from all VMs and adds an additional API:
```
SubscribeToEvents(ctx context.Context, pChainHeight uint64) (Message, uint64)
```

Which returns when an event (pending transactions or state sync done Message) is sent by the VM. The SubscribeToEvents receives additionally a P-chain height, which is used in the Warp precompile.

Unlike the toEngine pattern which only notifies the consumer a single time until a block is built, successive calls to `SubscribeToEvents` will return that transactions are pending as long as the mem-pool is not empty and a block can be built, otherwise the calls block.

The API change is needed to support consensus protocols which facilitate censorship resistance as part of their protocol and therefore need information about whether a block should be built or not on demand, in contrast to the curren toEngine pattern which notifies the consensus when new transactions arrive only once, and then block building may even fail. Having such an API on the VM side (and not implementing the logic in consensus) forces the VM to handle any side effects of block building (such as re-introducing transactions into the mempool).

## Why this should be merged

## How this works

## How this was tested

## Need to be documented in RELEASES.md?

StephenButtolph commented on chains/linearizable_vm.go:
```diff
@@ -33,11 +33,14 @@ type initializeOnLinearizeVM struct {
 	appSender    common.AppSender
 }
 
+func (vm *initializeOnLinearizeVM) SubscribeToEvents(ctx context.Context) common.Message {
+	return vm.vmToInitialize.SubscribeToEvents(ctx)
+}
```
Should we document that this is only allowed to be called after `Linearize`?

yacovm commented on snow/engine/common/subscription.go:
```diff
@@ -0,0 +1,270 @@
+
+type Subscriber interface {
+	// SubscribeToEvents blocks until either the given context is cancelled, or a message is returned.
+	SubscribeToEvents(ctx context.Context) Message
```
APIs are supposed to be simple and [synchronous](https://go.dev/talks/2013/bestpractices.slide#25). If I pass a channel, the only way of listening to this API is via a different goroutine. 

The way we call this method is via two code paths:

1. A goroutine in snowman constantly listens for events from this function and then invokes the snowman Engine's [Notify()](https://github.com/ava-labs/avalanchego/blob/c2ea301a95257da27f9694fc7d1482de48e837eb/snow/engine/snowman/engine.go#L442) API.
2. Simplex consensus runs a background goroutine which blocks until the VM tells it that the leader node (not the caller node, but a different node) needs to build a block.
3. Simplex consensus calls this method in a leader node to wait until a block needs to be built. This is to prevent from building an empty block. 

I will change its name to `WaitForEvent`.

yacovm commented:
> > Having a dedicated API allows consensus to obtain information about whether a block should be built on demand and therefore correlated to a specific round, while using the current toEngine pattern makes it impossible to correlate with a specific round, and this makes it impossible to incorporate with consensus.
> 
> couple q's clarifying your PR description
> 
> 1. By "correlate to a specific round" do you mean we can assume a next round has started once we receive a notification for a block to be built?
> 2. In simplex for each round we can either verify and accept the block or accept an empty block. In `avalanchego` does this mean we are just getting notified of new transactions? there isn't a notion of empty blocks so im trying to understand `toEngine` previous functionality.

1. No, I mean that we need to be certain that the API from the VM about the need to build a block is referring to the current round, and not for the previous round. Suppose you accept a block and then receive the notification from `toEngine` (because of its asynchronous nature). The notification can be about a transaction in the block you have just accepted. This is a problem, as it will be a false positive. We therefore need an API that will tell us "give me notifications starting from now"
2. The functionality of the `toEngine` is to wake up the snowman engine so it will build a new block. The problem is that in the current implementation, the `toEngine` won't give a notification if you haven't built a block. Therefore, it cannot be used to track a lazy / faulty leader as if you get notified by that channel but you're not the leader of the next round, you will never get notified again. So we essentially cannot use it for leader failover. ProposalVM relies on pre-computed time slots for when a leader should have built a block, but in Simplex we decided to only do a leader change if there are actual transactions pending in the mempool.

samliok commented:
> > > Having a dedicated API allows consensus to obtain information about whether a block should be built on demand and therefore correlated to a specific round, while using the current toEngine pattern makes it impossible to correlate with a specific round, and this makes it impossible to incorporate with consensus.
> > 
> > 
> > couple q's clarifying your PR description
> > 
> > 1. By "correlate to a specific round" do you mean we can assume a next round has started once we receive a notification for a block to be built?
> > 2. In simplex for each round we can either verify and accept the block or accept an empty block. In `avalanchego` does this mean we are just getting notified of new transactions? there isn't a notion of empty blocks so im trying to understand `toEngine` previous functionality.
> 
> 1. No, I mean that we need to be certain that the API from the VM about the need to build a block is referring to the current round, and not for the previous round. Suppose you accept a block and then receive the notification from `toEngine` (because of its asynchronous nature). The notification can be about a transaction in the block you have just accepted. This is a problem, as it will be a false positive. We therefore need an API that will tell us "give me notifications starting from now"
> 2. The functionality of the `toEngine` is to wake up the snowman engine so it will build a new block. The problem is that in the current implementation, the `toEngine` won't give a notification if you haven't built a block. Therefore, it cannot be used to track a lazy / faulty leader as if you get notified by that channel but you're not the leader of the next round, you will never get notified again. So we essentially cannot use it for leader failover. ProposalVM relies on pre-computed time slots for when a leader should have built a block, but in Simplex we decided to only do a leader change if there are actual transactions pending in the mempool.


makes sense, thanks for the context. To follow up on
> the `toEngine` won't give a notification if you haven't built a block.

does this mean we should add another message type to the `Message` [enum](https://github.com/yacovm/avalanchego/blob/e94c50f91a73b379aa65016e1feb080cd7a1cf67/snow/engine/common/message.go#L11) say `BlockShouldBeBuilt` message?

yacovm commented:
> > > > Having a dedicated API allows consensus to obtain information about whether a block should be built on demand and therefore correlated to a specific round, while using the current toEngine pattern makes it impossible to correlate with a specific round, and this makes it impossible to incorporate with consensus.
> > > 
> > > 
> > > couple q's clarifying your PR description
> > > 
> > > 1. By "correlate to a specific round" do you mean we can assume a next round has started once we receive a notification for a block to be built?
> > > 2. In simplex for each round we can either verify and accept the block or accept an empty block. In `avalanchego` does this mean we are just getting notified of new transactions? there isn't a notion of empty blocks so im trying to understand `toEngine` previous functionality.
> > 
> > 
> > 
> > 1. No, I mean that we need to be certain that the API from the VM about the need to build a block is referring to the current round, and not for the previous round. Suppose you accept a block and then receive the notification from `toEngine` (because of its asynchronous nature). The notification can be about a transaction in the block you have just accepted. This is a problem, as it will be a false positive. We therefore need an API that will tell us "give me notifications starting from now"
> > 2. The functionality of the `toEngine` is to wake up the snowman engine so it will build a new block. The problem is that in the current implementation, the `toEngine` won't give a notification if you haven't built a block. Therefore, it cannot be used to track a lazy / faulty leader as if you get notified by that channel but you're not the leader of the next round, you will never get notified again. So we essentially cannot use it for leader failover. ProposalVM relies on pre-computed time slots for when a leader should have built a block, but in Simplex we decided to only do a leader change if there are actual transactions pending in the mempool.
> 
> makes sense, thanks for the context. To follow up on
> 
> > the `toEngine` won't give a notification if you haven't built a block.
> 
> does this mean we should add another message type to the `Message` [enum](https://github.com/yacovm/avalanchego/blob/e94c50f91a73b379aa65016e1feb080cd7a1cf67/snow/engine/common/message.go#L11) say `BlockShouldBeBuilt` message? not in this pr(just asking if this is how it would be used for the simplex integration)

What I said applies to the current way things work. In my PR for coreth I changed the behavior, as we have the API I added and I removed the `toEngine` channel.

samliok reviewed (commented):
first pass

yacovm commented on snow/engine/common/subscription.go:
```diff
@@ -0,0 +1,270 @@
+
+func (ss *SimpleSubscriber) SubscribeToEvents(ctx context.Context) Message {
+	ss.lock.Lock()
+	defer ss.lock.Unlock()
```
When you execute the `ss.signal.Wait()` the lock is released. When you return from the execution, the lock is re-acquired.

samliok replied:
oh i see, i missed this initialization

ss.signal = *sync.NewCond(&ss.lock)

yacovm commented on snow/engine/common/notifier.go:
```diff
@@ -0,0 +1,89 @@
+		ctx := nf.getContext()
+
+		nf.Log.Debug("Subscribing to notifications")
+		msg := nf.Subscribe(ctx)
```
The only way to cancel this context, is if `Close()` is invoked. In such a case, we first close the `closeChan` and then cancel the context. Cancelling the context makes us return from `Subscribe` and we return due to the select discovering that the `closeChan` is closed. 

So, the only way which the context is cancelled, makes us return via the `closeChan` being closed, so we don't need to check if the context is cancelled.

yacovm commented on vms/proposervm/vm_test.go:
```diff
@@ -810,6 +810,19 @@ func TestExpiredBuildBlock(t *testing.T) {
+		case <-testEnd:
+			return 0
+		case event := <-events:
+			fmt.Println(">>>>>", event)
```
whoops, yes! thanks.

yacovm commented on vms/proposervm/vm_test.go:
```diff
@@ -871,24 +879,26 @@ func TestExpiredBuildBlock(t *testing.T) {
 		nil,
 		nil,
 	))
 	defer func() {
 		require.NoError(proVM.Shutdown(context.Background()))
 	}()
```
It is valid syntax but not the semantics we want. 

```

package main

import "fmt"

var n int

func h() {
	n = 3
}

func f() {
	defer fmt.Println(n)
	n = 1
}

func g() {
	defer func() {
		fmt.Println(n)
	}()
	n = 2
}

func main() {
	f() // prints 0
	g() // prints 2
}

```

yacovm commented on vms/proposervm/vm.go:
```diff
@@ -46,6 +46,12 @@ const (
 	innerBlkCacheSize = 64 * units.MiB
 )
 
+type SelfSubscriber interface {
```
No, it's because in Golang, interfaces should be declared where they're used, not where they're implemented

yacovm commented on chains/manager.go:
```diff
@@ -885,8 +879,8 @@ func (m *manager) createAvalancheChain(
 	// Asynchronously passes messages from the network to the consensus engine
 	h, err := handler.New(
 		ctx,
+		linearizableVM,
```
maybe but i think what I did also works 🤷‍♂️

yacovm commented on vms/proposervm/scheduler/scheduler.go:
```diff
@@ -73,21 +66,13 @@ waitloop:
-						zap.Stringer("messageString", msg),
-					)
-				}
+			case <-s.subscriptionDelayer.Forward(ctx):
```
The scheduler is a part of the proposerVM. The proposerVM sort of figures out when the node should propose and tells that to the scheduler.  In some sense, the scheduler throttles the inner VM's desire to build blocks and only builds blocks when the proposerVM's schedule allows it. 

The point of the proxy is to move the subscription from the inner VM to the scheduler.

>  I think this makes sense although, I'm not exactly sure why we'd want to delay building a block.

Because we want to ensure a single block proposer. Imagine that we have the inner VM telling the proposerVM to build blocks all the time, across all nodes, but we only want to do it in a single node at a time.

yacovm commented on snow/engine/common/subscription.go:
```diff
@@ -0,0 +1,278 @@
+				}
+			}
+
+			if sp.closed {
```
I don't think it matters. If we close this it's because we're shutting down, so no block is going to be built.
What comments did you have in mind?

samliok replied:
ah i didn't know that closing meant only shutting down. In that case I think it's fine as is

yacovm commented on snow/engine/common/subscription_test.go:
```diff
@@ -0,0 +1,137 @@
+			expectedEvent: StateSyncDone,
+		},
+		{
+			name: "Cancel wait",
```
all right I added this check.

yacovm commented on vms/proposervm/scheduler/scheduler.go:
```diff
@@ -26,27 +25,22 @@ type Scheduler interface {
 
-func New(log logging.Logger, toEngine chan<- common.Message) (Scheduler, chan<- common.Message) {
-	vmToEngine := make(chan common.Message, cap(toEngine))
+func New(subscription common.Subscription, logger logging.Logger) (Scheduler, *common.SubscriptionProxy) {
```
It doesn't support multiple subscribers. It says in the comments: 

> It assumes that there is only one subscriber at a time

for both the proxy and the simple subscriptions.

yacovm commented on snow/engine/common/subscription.go:
```diff
@@ -0,0 +1,278 @@
+		case <-ctx.Done():
+			return 0, nil
+		default:
+			sp.signal.Wait()
```
I added a check and a test

samliok commented on snow/engine/common/subscription.go:
```diff
@@ -0,0 +1,278 @@
+	signal  sync.Cond
+	running sync.WaitGroup
+
+	absorbedMsg *Message
```
i'd say pending makes sense more sense

samliok commented on vms/rpcchainvm/vm_client.go:
```diff
@@ -408,6 +406,16 @@ func (vm *VMClient) CreateHTTP2Handler(ctx context.Context) (http.Handler, error
+func (vm *VMClient) WaitForEvent(ctx context.Context) (common.Message, error) {
+	resp, err := vm.client.WaitForEvent(ctx, &emptypb.Empty{})
+	if err != nil {
+		// TODO: Correctly return this error to the caller.
```
is this for a future pr?

StephenButtolph commented on snow/engine/common/notifier.go:
```diff
@@ -0,0 +1,107 @@
+		}
+		nf.Log.Debug("Received notification", zap.Stringer("msg", msg))
+
+		nf.cancelContext()
```
The implementation behind `Notifier.Notify` is:
```go
func (h *handler) Notify(_ context.Context, msg common.Message) error {
	h.msgFromVMChan <- msg
	return nil
}
```
So the current code doesn't care that the context is cancelled here.

StephenButtolph commented on utils/lock/cond_test.go:
```diff
@@ -0,0 +1,166 @@
+	cancelled, cancel := context.WithCancel(context.Background())
+	cancel()
+
+	log := func(args ...any) {
+		t.Log(args...)
+	}
```
It appears that this function was moved to fix a data-race (based on the commit message).
1. Aren't we using the wrong logger now? The sub-tests are now logging into the parent test.
2. Why does using the parent testing struct not incur the same data-race as the child testing struct?
3. I'm surprised that the stdlib logger doesn't support concurrent logs... It isn't implemented on the interface...

yacovm replied:
It actually supports concurrent calls, but the goroutines need to finish and not remain running once the test ends.

Fixed [here](https://github.com/ava-labs/avalanchego/pull/3999/commits/aa40197de464a9646a64b79fa21ff08b7d3b8db7#diff-21d4aef882132c8ba89d616ef779c42596a30aadefae9d8b66861b984ae8975fR112-R130).

And I don't know what I was thinking with the log function. I think I added a lock and somehow removed it by mistake.

StephenButtolph commented on vms/proposervm/vm.go:
```diff
@@ -315,13 +300,83 @@ func (vm *VM) SetPreference(ctx context.Context, preferred ids.ID) error {
+		// If we are pre-fork or haven't finished bootstrapping yet, we should
+		// directly forward the inner VM's events.
+		if !shouldWait {
+			vm.ctx.Log.Debug("Waiting for inner VM event (pre-fork)")
```
This code path is also used for state-sync, not just when the VM is pre-fork.

yacovm replied:
[Fixed](https://github.com/ava-labs/avalanchego/pull/3999/commits/aa40197de464a9646a64b79fa21ff08b7d3b8db7#diff-83207c67653d884434d01f4c9ad5255cdda73dabef39063346ac96f4a039530fR331-R337).

StephenButtolph commented on vms/proposervm/vm.go:
```diff
@@ -315,13 +300,83 @@ func (vm *VM) SetPreference(ctx context.Context, preferred ids.ID) error {
+
+		duration := time.Until(timeToBuild)
+		if duration <= 0 {
+			vm.ctx.Log.Debug("Building a block immediately")
```
immediately kind of feels like this is non-blocking. This statement may be logged after the `WaitForEvent` flow has been blocking for awhile.

yacovm replied:
[Fixed](https://github.com/ava-labs/avalanchego/pull/3999/commits/aa40197de464a9646a64b79fa21ff08b7d3b8db7#diff-83207c67653d884434d01f4c9ad5255cdda73dabef39063346ac96f4a039530fR331-R337).

StephenButtolph commented on vms/proposervm/vm_test.go:
```diff
@@ -810,6 +809,18 @@ func TestExpiredBuildBlock(t *testing.T) {
 		}
 	}
 
+	testEnd := make(chan struct{})
+
+	events := make(chan common.Message, 1)
+	coreVM.WaitForEventF = func(context.Context) (common.Message, error) {
+		select {
+		case <-testEnd:
+			return 0, nil
+		case event := <-events:
+			return event, nil
+		}
+	}
```
It feels weird to use `testEnd` like this. This isn't called in another goroutine, so what are we trying to exit?

If we want to keep an implementation that supports cancelation, I feel like we should use the passed in context instead.

yacovm replied:
Fixed [here](https://github.com/ava-labs/avalanchego/pull/3999/commits/aa40197de464a9646a64b79fa21ff08b7d3b8db7#diff-0a19ad2bab665d2f2f629cb3a692b258012a9416ebe786715959abfd915a575dR813).

StephenButtolph commented on vms/proposervm/vm_test.go:
```diff
@@ -2000,6 +1995,8 @@ func TestVM_VerifyBlockWithContext(t *testing.T) {
 
 	// Create a VM
 	innerVM := blockmock.NewChainVM(ctrl)
+	innerVM.EXPECT().WaitForEvent(gomock.Any()).Return(common.Message(0), nil).AnyTimes()
```
Should we be returning `common.PendingTxs` here? `0, nil` is really never valid to be returned.

StephenButtolph commented on vms/txs/mempool/mempool_test.go:
```diff
@@ -297,3 +301,37 @@ func TestBlockBuilderMaxMempoolSizeHandling(t *testing.T) {
+func TestWaitForEventWithTx(t *testing.T) {
+	require := require.New(t)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	var addErr atomic.Value
+
+	m := newMempool()
+	go func() {
+		defer wg.Done()
+		tx := newTx(0, 32)
+		if err := m.Add(tx); err != nil {
+			addErr.Store(err)
+		}
+	}()
+
+	msg, err := m.WaitForEvent(context.Background())
+	require.NoError(err)
+	require.Equal(common.PendingTxs, msg)
+	wg.Wait()
+	require.Nil(addErr.Load())
```
```suggestion
	m := newMempool()
	errs := make(chan error)
	go func() {
		tx := newTx(0, 32)
		errs <- m.Add(tx)
	}()

	msg, err := m.WaitForEvent(context.Background())
	require.NoError(err)
	require.Equal(common.PendingTxs, msg)
	require.NoError(<-errs)
```

yacovm replied:
Applied [here](https://github.com/ava-labs/avalanchego/pull/3999/commits/3c959c1ef74b5287cab18a278699a98e2b835831).

StephenButtolph commented on chains/notifier.go:
```diff
@@ -0,0 +1,33 @@
+	"github.com/ava-labs/avalanchego/snow/engine/snowman/block"
+)
+
+type ChangeNotifier struct {
+	block.ChainVM
+	OnChange func()
+}
```
While this method of implementing a VM to intercept the calls is nice... I don't think this correctly handles the optional interfaces that VMs may implement.

`ChangeNotifier` will only implement `block.ChainVM` even if the `block.ChainVM` that it wraps implements other optional VM interfaces (such as state sync and batched operations).

As an example, I don't think that the C-chain will be able to state-sync with this wrapping.

StephenButtolph commented on chains/manager.go:
```diff
@@ -1292,6 +1301,16 @@ func (m *manager) createSnowmanChain(
+
+	cn.OnChange = nf.PreferenceOrStateChanged
+
+	defer nf.Start()
```
Where do we close the `nf`?

StephenButtolph commented on chains/manager.go:
```diff
@@ -900,6 +900,16 @@ func (m *manager) createAvalancheChain(
+
+	cn.OnChange = nf.PreferenceOrStateChanged
+
+	defer nf.Start()
```
Similar question about where this is closed

yacovm replied:
Fixed [here](https://github.com/ava-labs/avalanchego/pull/3999/commits/7510692702967fac32c55c5a7d81b04d8b737a49).

StephenButtolph commented on chains/linearizable_vm.go:
```diff
@@ -28,26 +28,41 @@ type initializeOnLinearizeVM struct {
 
 func (vm *initializeOnLinearizeVM) Linearize(ctx context.Context, stopVertexID ids.ID) error {
 	vm.vmToLinearize.stopVertexID = stopVertexID
+	select {
+	case <-vm.waitForLinearize:
+	default:
+		defer func() {
+			close(vm.waitForLinearize)
+		}()
+	}
```
While this does work to ensure that the channel is only closed once... It feels odd. Would it be cleaner to use a `sync.Once`? Or honestly even panic if we call `Linearize` multiple times.

yacovm commented on snow/engine/common/notifier.go:
```diff
@@ -0,0 +1,127 @@
+	default:
+	}
+
+	if err := nf.Engine.Notify(context.Background(), msg); err != nil {
+		nf.Log.Debug("Failed notifying engine", zap.Error(err))
+		return
+	}
```
If I pass `ctx` then once I'm interrupted by a preference change, I will lose the message I am trying to pass to the engine, no?

StephenButtolph replied:
That's correct. Why is that an issue? The context could have been cancelled when we were calling `WaitForEvent` anyways.

yacovm replied:
You're right, [fixed](https://github.com/ava-labs/avalanchego/pull/3999/commits/bb0ac27a6a4d7a1f3d079d858b03702664f2dae8).

yacovm commented on snow/engine/common/notifier.go:
```diff
@@ -0,0 +1,127 @@
+		nf.lock.Unlock()
+	default:
+		close(nf.closeChan)
+		nf.lock.Unlock()
```
Yes good idea, fixed [here](https://github.com/ava-labs/avalanchego/pull/3999/commits/ce75328c44161800ff6acb887c07af9eaf7024e6).

yacovm commented on snow/engine/common/notifier.go:
```diff
@@ -0,0 +1,127 @@
+	nf.running.Add(1)
+	nf.closeChan = make(chan struct{})
+	go nf.run()
+}
```
Fixed [here](https://github.com/ava-labs/avalanchego/pull/3999/commits/aaef80db3c26ffd9d87d867119b57872dd0d8c04).

yacovm commented on snow/engine/enginetest/vm.go:
```diff
@@ -59,6 +59,14 @@ type VM struct {
+	if vm.WaitForEventF != nil {
+		return vm.WaitForEventF(context.Background())
+	}
+	return common.Message(0), nil
```
Fixed [here](https://github.com/ava-labs/avalanchego/pull/3999/commits/947c657186610bd761dd73ceb256af79303c4872#diff-5858520123cf8dab7b94b4b0ebbe1c43f083015e9406cb1c8b5060c7964d3768R65-R69).

yacovm commented on snow/engine/snowman/engine_test.go:
```diff
@@ -3234,7 +3231,10 @@ func TestEngineAbortQueryWhenInPartition(t *testing.T) {
+	// The lock is needed because the engine logs into the log asynchronously
+	conf.Ctx.Lock.Lock()
 	require.Contains(buff.String(), errInsufficientStake)
+	conf.Ctx.Lock.Unlock()
```
not anymore, [fixed](https://github.com/ava-labs/avalanchego/pull/3999/commits/947c657186610bd761dd73ceb256af79303c4872#diff-cb027abbde9258d1a8393868fce1a35f0b167fb125185b93ce2e9f2b016e2a37L3234-L3237).

yacovm commented on snow/networking/handler/handler.go:
```diff
@@ -278,6 +279,11 @@ func (h *handler) Start(ctx context.Context, recoverPanic bool) {
+func (h *handler) Notify(_ context.Context, msg common.Message) error {
+	h.msgFromVMChan <- msg
+	return nil
+}
```
Yep, [fixed](https://github.com/ava-labs/avalanchego/pull/3999/commits/bb0ac27a6a4d7a1f3d079d858b03702664f2dae8#diff-1b154e2d9b5b0909a66199417df6ff79ee9a18e83075b39ef2acf3634558342eR282-R285).

yacovm commented on snow/engine/common/notifier.go:
```diff
@@ -0,0 +1,133 @@
+// and its role is to signal the NotificationForwarder to stop its current subscription and re-subscribe.
+// This is needed in case a block has been accepted that changes when a VM considers the need to build a block.
+// In order for the subscription to be correlated to the latest data, it needs to be retried.
+func (nf *NotificationForwarder) PreferenceOrStateChanged() {
```
Renamed it [here](https://github.com/ava-labs/avalanchego/pull/3999/commits/a07900daa6f32fa9a06d48142fcc6a1322b7e779).

yacovm commented on snow/engine/common/notifier.go:
```diff
@@ -0,0 +1,133 @@
+			return
+		default:
+		}
+	}
```
[Fixed](https://github.com/ava-labs/avalanchego/pull/3999/commits/c77b679c6e410f12e820fe0bdc7c3a907801808a#diff-54a0e1a7b49ba8be40d7830923bb1c962d32e15c7a98a5b7460c957b0e224f20R62).

yacovm commented on snow/engine/common/notifier.go:
```diff
@@ -0,0 +1,133 @@
+
+	if nf.abortContext != nil {
+		nf.abortContext()
+		nf.abortContext = nil
```
[Fixed](https://github.com/ava-labs/avalanchego/pull/3999/commits/c77b679c6e410f12e820fe0bdc7c3a907801808a#diff-54a0e1a7b49ba8be40d7830923bb1c962d32e15c7a98a5b7460c957b0e224f20L109).

yacovm commented on snow/engine/common/notifier.go:
```diff
@@ -0,0 +1,133 @@
+	case <-nf.execCtx.Done():
+		return
+	case <-ctx.Done():
+	}
```
That's true. [Fixed](https://github.com/ava-labs/avalanchego/pull/3999/commits/c77b679c6e410f12e820fe0bdc7c3a907801808a#diff-54a0e1a7b49ba8be40d7830923bb1c962d32e15c7a98a5b7460c957b0e224f20L91-L95).

yacovm commented on snow/engine/common/vm.go:
```diff
@@ -81,4 +80,7 @@ type VM interface {
 	// avalanchego api server.
 	CreateHTTP2Handler(ctx context.Context) (http.Handler, error)
+
+	// WaitForEvent blocks until either the given context is cancelled, or a message is returned.
```
> All WaitForEvent calls are guaranteed to have returned prior to Shutdown being called.

I am not sure I understand why we would want to guarantee such a thing. 
What's the harm in making any call to `WaitForEvent` be a no-op after shutdown has been called?

yacovm commented on snow/engine/enginetest/vm.go:
```diff
@@ -59,6 +59,14 @@ type VM struct {
+	if vm.WaitForEventF != nil {
+		return vm.WaitForEventF(context.Background())
+	}
+	return 0, ctx.Err()
```
True, [fixed](https://github.com/ava-labs/avalanchego/pull/3999/commits/c77b679c6e410f12e820fe0bdc7c3a907801808a#diff-5858520123cf8dab7b94b4b0ebbe1c43f083015e9406cb1c8b5060c7964d3768R66-R69).

yacovm commented on snow/engine/enginetest/vm.go:
```diff
@@ -59,6 +59,14 @@ type VM struct {
+
+func (vm *VM) WaitForEvent(ctx context.Context) (common.Message, error) {
+	if vm.WaitForEventF != nil {
+		return vm.WaitForEventF(context.Background())
```
Thanks, [fixed](https://github.com/ava-labs/avalanchego/pull/3999/commits/c77b679c6e410f12e820fe0bdc7c3a907801808a#diff-5858520123cf8dab7b94b4b0ebbe1c43f083015e9406cb1c8b5060c7964d3768R67).

yacovm commented on chains/notifier.go:
```diff
@@ -0,0 +1,83 @@
+	"github.com/ava-labs/avalanchego/snow/consensus/snowman"
+	"github.com/ava-labs/avalanchego/snow/engine/snowman/block"
+)
+
```
[Added](https://github.com/ava-labs/avalanchego/pull/3999/commits/c77b679c6e410f12e820fe0bdc7c3a907801808a#diff-8be530da17b2a267edd0b19a823fa9de92773a962f6a4b795bfa292e2a9859c9R15-R20).

yacovm commented on snow/engine/common/notifier_test.go:
```diff
@@ -0,0 +1,164 @@
+	}, time.Minute, 10*time.Millisecond)
+
+	require.Never(t, func() bool {
+		return atomic.LoadUint32(&notifiedCount) == uint32(2)
```
[Fixed](https://github.com/ava-labs/avalanchego/pull/3999/commits/c77b679c6e410f12e820fe0bdc7c3a907801808a#diff-525d9d6b02531ca187cf35424072e3648866bbf2bda773a99cab0bfc451a8346).

yacovm commented on snow/engine/common/notifier_test.go:
```diff
@@ -0,0 +1,164 @@
+	nf.PreferenceOrStateChanged()
+
+	require.Eventually(t, func() bool {
+		return atomic.LoadUint32(&notifiedCount) == uint32(2)
```
[Fixed](https://github.com/ava-labs/avalanchego/pull/3999/commits/c77b679c6e410f12e820fe0bdc7c3a907801808a#diff-525d9d6b02531ca187cf35424072e3648866bbf2bda773a99cab0bfc451a8346).

StephenButtolph commented on snow/networking/router/chain_router_test.go:
```diff
@@ -1603,3 +1625,14 @@ func TestHandleSimplexMessage(t *testing.T) {
 	require.True(t, receivedMsg)
 }
+
+func createSubscriberAndChangeNotifier() (*block.ChangeNotifier, common.Subscription) {
+	var cn block.ChangeNotifier
+
+	subscription := func(ctx context.Context) (common.Message, error) {
+		<-ctx.Done()
+		return common.Message(0), ctx.Err()
+	}
+
+	return &cn, subscription
+}
```
Why do we have `createSubscriberAndChangeNotifier`? The returned `subscription` doesn't have any state related to the `ChangeNotifier` (actually any state at all).

It seems like the callers of this should just define `var cn block.ChangeNotifier` themselves and reference a single static `noopSubscription` function.
```suggestion
func noopSubscription(ctx context.Context) (common.Message, error)
	<-ctx.Done()
	return common.Message(0), ctx.Err()
}
```

StephenButtolph commented on snow/networking/sender/sender_test.go:
```diff
@@ -1308,3 +1318,14 @@ func TestSender_Single_Request(t *testing.T) {
 	}
 }
+
+func createSubscriberAndChangeNotifier() (*block.ChangeNotifier, common.Subscription) {
+	var cn block.ChangeNotifier
+
+	subscription := func(ctx context.Context) (common.Message, error) {
+		<-ctx.Done()
+		return common.Message(0), ctx.Err()
+	}
+
+	return &cn, subscription
+}
```
Same comment as in the router. Perhaps the noop subscription could be some test helper somewhere, but tbh it's so simple that I'm not concerned about the duplication.

StephenButtolph commented on snow/networking/handler/handler_test.go:
```diff
@@ -653,3 +682,20 @@ func TestHandlerStartError(t *testing.T) {
 	require.NoError(err)
 }
+
+func createSubscriberAndChangeNotifier() (*block.ChangeNotifier, common.Subscription, chan<- common.Message) {
+	var cn block.ChangeNotifier
+
+	messages := make(chan common.Message, 1)
+
+	subscription := func(ctx context.Context) (common.Message, error) {
+		select {
+		case msg := <-messages:
+			return msg, nil
+		case <-ctx.Done():
+			return common.Message(0), ctx.Err()
+		}
+	}
+
+	return &cn, subscription, messages
+}
```
Similar comment as to the router, it doesn't seem like the `ChangeNotifier` is related to the `subscription` or the `messages` channel from this function. Can we remove the `ChangeNotifier` from this function?

StephenButtolph commented on snow/networking/handler/handler.go:
```diff
@@ -46,11 +47,14 @@ var (
 type Handler interface {
 	health.Checker
 
+	Notify(_ context.Context, msg common.Message) error
```
Does `Notify` need to be added to the `Handler` interface? Now that we are creating the `NotificationForwarder` internally, I think we pass in the struct rather than an interface, so it feels like this isn't needed.

StephenButtolph commented on snow/engine/snowman/syncer/state_syncer.go:
```diff
@@ -577,7 +577,7 @@ func (ss *stateSyncer) sendGetAcceptedStateSummaries(ctx context.Context) {
 func (ss *stateSyncer) Notify(ctx context.Context, msg common.Message) error {
 	if msg != common.StateSyncDone {
-		ss.Ctx.Log.Warn("received an unexpected message from the VM",
+		ss.Ctx.Log.Info("received an unexpected message from the VM",
```
Why was this changed from `Warn` to `Info`? This is still unexpected behavior right?

StephenButtolph commented on snow/engine/snowman/engine.go:
```diff
@@ -450,7 +450,7 @@ func (e *Engine) Notify(ctx context.Context, msg common.Message) error {
 		return nil
 	default:
-		e.Ctx.Log.Warn("received an unexpected message from the VM",
+		e.Ctx.Log.Info("received an unexpected message from the VM",
```
Why was this changed from `Warn` to `Info`? This is still unexpected behavior right?

StephenButtolph commented on snow/engine/snowman/bootstrap/bootstrapper.go:
```diff
@@ -755,7 +755,7 @@ func (b *Bootstrapper) restartBootstrapping(ctx context.Context) error {
 func (b *Bootstrapper) Notify(_ context.Context, msg common.Message) error {
 	if msg != common.StateSyncDone {
-		b.Ctx.Log.Warn("received an unexpected message from the VM",
+		b.Ctx.Log.Info("received an unexpected message from the VM",
```
Why was this changed from `Warn` to `Info`? This is still unexpected behavior right?

StephenButtolph commented on snow/engine/common/notifier.go:
```diff
@@ -0,0 +1,123 @@
+func (nf *NotificationForwarder) Close() {
+	defer nf.executing.Wait()
+
+	nf.lock.Lock()
+	defer nf.lock.Unlock()
+
```
This lock isn't needed.
```suggestion
```

StephenButtolph commented on chains/manager.go:
```diff
@@ -266,7 +266,7 @@ type manager struct {
 	// Key: Chain's ID
 	// Value: The chain
-	chains map[ids.ID]handler.Handler
+	chains map[ids.ID]*chain
```
I think this diff can be removed.

StephenButtolph commented on snow/networking/handler/handler_test.go:
```diff
@@ -653,3 +682,18 @@ func TestHandlerStartError(t *testing.T) {
 	require.NoError(err)
 }
+
+func createSubscriberAndChangeNotifier() (common.Subscription, chan<- common.Message) {
```
```suggestion
func createSubscriber() (common.Subscription, chan<- common.Message) {
```

StephenButtolph commented on vms/platformvm/vm_test.go:
```diff
@@ -2189,3 +2177,14 @@ func TestPruneMempool(t *testing.T) {
 	require.True(ok)
 }
+
+func createSubscriberAndChangeNotifier() (*smblock.ChangeNotifier, common.Subscription) {
+	var cn smblock.ChangeNotifier
+
+	subscription := func(ctx context.Context) (common.Message, error) {
+		<-ctx.Done()
+		return common.Message(0), ctx.Err()
+	}
+
+	return &cn, subscription
+}
```
Sorry I forgot to comment on this function in the prior review as well. This guy should be cleaned up with the same `noopSubscription`

StephenButtolph commented on chains/manager.go:
```diff
@@ -811,19 +806,25 @@ func (m *manager) createAvalancheChain(
 	}
 
+	cn := &block.ChangeNotifier{
+		ChainVM: vmWrappingProposerVM,
```
`OnChange` is populated by `handler.New`. This initialization flow is a bit gross, but it allows us to stitch together components without having a circular dependency.

yacovm replied:
unfortunately there is no way to masquerade as a VM inside the handler because we need to do it in the `manager.go`

yacovm commented on snow/engine/snowman/block/notifier.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+func (cn *ChangeNotifier) SetState(ctx context.Context, state snow.State) error {
+	defer cn.OnChange()
```
we only do `SetState` twice in our lifecycle - when we move from state sync to bootstrapping and then to normal operation.

StephenButtolph replied:
re: `unnecessarily resubscribing`. We do currently call `OnChange` every time `SetPreference` is called. In the consensus code, we don't restrict our usage of `SetPreference` to when the preference has actually changed, so we are going to call `SetPreference` more than our preference actually changes.

If we wanted to avoid this, we could either change consensus to keep track of the last updated preference, or we could implement that inside the `ChangeNotifier`.

I had been planning on looking into the metrics after we canaried this out to see if this was an important optimization, but if we wanted to keep parity with the current code, we could implement this in this PR.

To clarify, this optimization was previously implemented in the proposervm's and platformvm's `SetPreference` implementations.
